product_optional_product_quantity/models/product_optional_line.py

```diff
@@ -0,0 +1,57 @@
+# Copyright 2024 Cetmix OÜ
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import _, api, fields, models
+
+
+class ProductOptionalLine(models.Model):
+    _name = "product.optional.line"
+    _description = "Optional Product"
+
+    product_tmpl_id = fields.Many2one(
+        comodel_name="product.template",
+        string="Related Product",
+        required=True,
+    )
+    company_id = fields.Many2one(related="product_tmpl_id.company_id")
+    optional_product_tmpl_id = fields.Many2one(
+        comodel_name="product.template",
+        domain="[('company_id', 'in', [company_id, False])]",
+        string="Product",
+        required=True,
+    )
+    quantity = fields.Float(
+        default=1.0,
+        required=True,
+    )
+
+    @api.constrains("product_tmpl_id", "optional_product_tmpl_id")
```

Comment:
I think you can also add a domain in both fields, for example for **optional_product_tmpl_id**:

`domain="[('id', '!=', product_tmpl_id)]"`

This way, the user will not be able to select the same product.